sdk/pinocchio/src/cpi.rs

```diff
@@ -13,148 +13,253 @@ use crate::{
+    // Check that the number of accounts provided is not less than
+    // the number of accounts expected by the instruction.
+    if account_infos.len() < instruction.accounts.len() {
+        return Err(ProgramError::NotEnoughAccountKeys);
     }
 
     const UNINIT: MaybeUninit<Account> = MaybeUninit::<Account>::uninit();
-    let mut accounts = [UNINIT; MAX_CPI_ACCOUNTS];
-    let mut len = 0;
-
-    for (account_info, account_meta) in account_infos.iter().zip(instruction.accounts.iter()) {
-        if account_info.key() != account_meta.pubkey {
-            return Err(ProgramError::InvalidArgument);
-        }
-
-        let state = if account_meta.is_writable {
-            BorrowState::Borrowed
-        } else {
-            BorrowState::MutablyBorrowed
-        };
-
-        if account_info.is_borrowed(state) {
-            return Err(ProgramError::AccountBorrowFailed);
-        }
-
-        // SAFETY: The number of accounts has been validated to be less than
-        // `MAX_CPI_ACCOUNTS`.
-        unsafe {
-            accounts
-                .get_unchecked_mut(len)
-                .write(Account::from(*account_info));
-        }
-
-        len += 1;
-    }
-    // SAFETY: The accounts have been validated.
+    let mut accounts = [UNINIT; MAX_ACCOUNTS];
+
+    account_infos
+        .iter()
+        .zip(instruction.accounts.iter())
+        .zip(accounts.iter_mut())
+        .try_for_each(|((account_info, account_meta), account)| {
+            // In order to check whether the borrow state is compatible
+            // with the invocation, we need to check that we have the
+            // correct account info and meta pair.
+            if account_info.key() != account_meta.pubkey {
+                return Err(ProgramError::InvalidArgument);
+            }
+
+            // Determines the borrow state that would be invalid according
+            // to their mutability on the instruction.
+            let invalid_state = if account_meta.is_writable {
+                // If the account is required to be writable, it cannot
+                //  be currently borrowed.
+                BorrowState::Borrowed
+            } else {
+                // If the account is required to be read-only, it cannot
+                // be currently mutably borrowed.
+                BorrowState::MutablyBorrowed
+            };
+
+            if account_info.is_borrowed(invalid_state) {
+                return Err(ProgramError::AccountBorrowFailed);
+            }
+
+            account.write(Account::from(*account_info));
+
+            Ok(())
+        })?;
+
+    // SAFETY: At this point it is guaranteed that account infos are borrowable
+    // according to their mutability on the instruction.
     unsafe {
         invoke_signed_unchecked(
             instruction,
-            core::slice::from_raw_parts(accounts.as_ptr() as _, len),
+            from_raw_parts(accounts.as_ptr() as _, instruction.accounts.len()),
```

Comment:
I think this function is missing a check, that ensures that `MAX_ACCOUNTS` is greater or equal to the `instruction.accounts.len()`.
If I call
```rust
invoke_signed_with_bounds<1>(
    /* instruction that takes 2 accounts */,
    &[ /* two AccountInfo reference */],
    &[]
)
```
this `invoke_signed_unchecked()` would be accessing past the allocated `accounts` array, no?